h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java

```diff
@@ -329,15 +338,53 @@ public void map(Chunk[] chks, NewChunk[] cress) {
           if (clf.vec().isString())
             for (int i = 0; i < clf._len; i++)
               cres.addNum(str_op(clf.atStr(lfstr, i), crt.atStr(rtstr, i)));
-          else
+          else if (categoricalOK() && clf.vec().isCategorical() && crt.vec().isCategorical()) {
+            // The vec with longer domain is iterated over due to categorical mapping
+            if (clf.vec().domain().length > crt.vec().domain().length) {
+              for (int i = 0; i < clf._len; i++) {
+                cres.addCategorical((int) op(clf.atd(i), alignedCategoricals[c][(int) crt.atd(i)]));
+              }
+            } else {
+              for (int i = 0; i < clf._len; i++) {
+                cres.addCategorical((int) op(alignedCategoricals[c][(int) clf.atd(i)], crt.atd(i)));
+              }
+            }
+          } else {
             for (int i = 0; i < clf._len; i++)
               cres.addNum(op(clf.atd(i), crt.atd(i)));
+          }
         }
       }
     }.doAll(lf.numCols(), Vec.T_NUM, new Frame(lf).add(rt)).outputFrame(lf._names, null);
     return cleanCategorical(lf, res); // Cleanup categorical misuse
   }
 
+  /**
+   * Produces a mapping array with indexes of the smaller pointing to the larger domain.
+   *
+   * @param sourceDomain Domain to originally map from
+   * @param targetDomain Domain to originally map to
+   * @return Cross-domain mapping as an array of primitive integers
+   */
+  private int[] alignCategoricals(String[] sourceDomain, String[] targetDomain) {
```

Comment:
Having domains passed as method arguments ensures those are copies and original domain arrays are not affected while sorting.